File: most_frequent_subtree_sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findFrequentTreeSum(self, root: Optional[TreeNode]) -> List[int]:
        freq = {}

        def dfs(node):
            if not node:
                return 0
            
            # calculate the subtree sum recursively
            left_sum = dfs(node.left)
            right_sum = dfs(node.right)

            subtree_sum = left_sum + right_sum + node.val
            # increment the frequency of the subtree sum
            freq[subtree_sum] = 1 + freq.get(subtree_sum, 0)
            
            return subtree_sum

        dfs(root)

        max_freq = max(freq.values())
        return [s for s, f in freq.items() if f == max_freq]

File: most_frequent_subtree_sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int[] findFrequentTreeSum(TreeNode root) {
        Map<Integer, Integer> freq = new HashMap<>();
        dfs(root, freq);
        
        int maxFreq = Collections.max(freq.values());
        return freq.entrySet().stream()
            .filter(entry -> entry.getValue() == maxFreq)
            .mapToInt(entry -> entry.getKey())
            .toArray();        
    }

    private int dfs(TreeNode node, Map<Integer, Integer> freq) {
        if (node == null) {
            return 0;
        }
        
        // calculate the subtree sum recursively
        int leftSum = dfs(node.left, freq);
        int rightSum = dfs(node.right, freq);
        int subtreeSum = leftSum + rightSum + node.val;
        
        // increment the frequency of the subtree sum
        freq.put(subtreeSum, freq.getOrDefault(subtreeSum, 0) + 1);
        
        return subtreeSum;
    }
}
